rust/common-graph/src/processor.rs

```diff
@@ -0,0 +1,282 @@
+    pub fn inputs(&self) -> &[(&'a str, Option<V>)] {
+        &self.inputs[..]
+    }
+
+    /// Returns the map of port name and value, representing
+    /// outgoing data. The data is mutable and able to be written
+    /// to.
+    pub fn outputs(&'a mut self) -> &mut [(&'a str, &mut Option<V>)] {
+        &mut self.outputs[..]
+    }
+}
+
+/// An iterating processor for [PortGraph].
+///
+/// A [GraphProcessor] takes inputs, and feeds them into an immutable [PortGraph]
+/// network. Using [GraphProcessor::next], once a node has all of its
+/// inputs fulfilled, the processor yields a node and its I/O data,
+/// where its output can be written. The node's output data propagates
+/// through the network, and the next node with complete input yields,
+/// and so forth, until all nodes have been processed.
+///
+/// A [GraphProcessor] is similar to an [Iterator],
+/// using the [lending iterator] pattern instead to "lend"
+/// mutable references with a lifetime valid until the
+/// subsequent `next()` call.
+///
+/// [lending iterator]: <https://docs.rs/lending-iterator>
+pub struct GraphProcessor<'a, T, V>
+where
+    T: Debug + 'a,
+    V: Clone + 'a,
+    Self: 'a,
+{
+    graph: &'a PortGraph<T>,
+    current_index: Option<usize>,
+    storage: GraphStorage<'a, V>,
+    queue: VecDeque<usize>,
+    output: Option<Vec<(String, Option<V>)>>,
+}
+
+impl<'a, T, V> GraphProcessor<'a, T, V>
+where
+    T: Debug + 'a,
+    V: Clone + 'a,
+    Self: 'a,
+{
+    /// Creates a new [GraphProcessor].
+    pub(crate) fn new<I>(graph: &'a PortGraph<T>, input: I) -> Result<Self>
+    where
+        I: IntoIterator<Item = (&'a str, V)>,
+    {
+        let mut queue = VecDeque::default();
+
+        // Seed the storage with port keys and queue up any
+        // input-less nodes first after the root node.
+        let mut storage = GraphStorage::from_iter(graph.nodes().iter().map(|node| {
+            if !node.is_root() && node.inputs().is_empty() {
+                queue.push_back(node.index())
+            }
+            return (
+                node.inputs().iter().map(|(key, _)| key.as_str()),
+                node.outputs().iter().map(|(key, _)| key.as_str()),
+            );
+        }));
+
+        // Seed input data on root node outputs
+        for (port, value) in input {
+            storage.set(0, port, value, PortType::Output)?;
+        }
+
+        Ok(Self {
+            graph,
+            storage,
+            current_index: Some(0),
+            queue,
+            output: None,
+        })
+    }
+}
+
+/// The lending iterator type trait implemented by [GraphProcessor].
+pub trait Processor {
+    /// Type of item to yield on `next()`.
+    type Item<'a>
+    where
+        Self: 'a;
+
+    /// Type of finalized output.
+    type Output<'out>
+    where
+        Self: 'out;
+
+    /// Yield the next node to process. If `None`,
+    /// there are no more items to process.
+    fn next(&mut self) -> Option<Self::Item<'_>>;
```

Comment:
`Option<Item>` to be an Iterator-like, but because we perform fallible processing between steps, maybe `Result<Option<Item>>` is more appropriate (and would lead to nicer usage, see `fixtures::Op::solve`)